drivers/SmartThings/jbl/src/lunchbox/rest.lua

```diff
@@ -0,0 +1,403 @@
+RestClient.__index = RestClient
+
+function RestClient.one_shot_get(full_url, additional_headers, socket_builder)
+  local url_table = utils.force_url_table(full_url)
+
+  return RestClient.new(url_table.scheme .. "://" .. url_table.authority, socket_builder):get(
+           url_table.path, additional_headers
+         )
+end
+
+function RestClient.one_shot_post(full_url, body, additional_headers, socket_builder)
+  local url_table = utils.force_url_table(full_url)
+
+  return RestClient.new(url_table.scheme .. "://" .. url_table.host, socket_builder):post(
+           url_table.path, body, additional_headers
+         )
+end
+
+function RestClient:update_base_url(new_url)
+  if self.socket ~= nil then
+    self.socket:close()
+    self.socket = nil
+  end
+
+  self.base_url = utils.force_url_table(new_url)
+end
+
+function RestClient:get(path, additional_headers, retry_fn)
+  local request = Request.new("GET", path, nil):add_header(
+                    "user-agent", "smartthings-lua-edge-driver"
+                  ):add_header("host", string.format("%s", self.base_url.host)):add_header(
+                    "connection", "keep-alive"
+                  )
+
+  if additional_headers ~= nil and type(additional_headers) == "table" then
+    for k, v in pairs(additional_headers) do request = request:add_header(k, v) end
+  end
+
+  return execute_request(self, request, retry_fn)
+end
+
+function RestClient:post(path, body_string, additional_headers, retry_fn)
+  local request = Request.new("POST", path, nil):add_header(
+                    "user-agent", "smartthings-lua-edge-driver"
+                  ):add_header("host", string.format("%s", self.base_url.host)):add_header(
+                    "connection", "keep-alive"
+                  )
+
+  if additional_headers ~= nil and type(additional_headers) == "table" then
+    for k, v in pairs(additional_headers) do request = request:add_header(k, v) end
+  end
+
+  request = request:append_body(body_string)
+
+  return execute_request(self, request, retry_fn)
+end
+
+function RestClient:put(path, body_string, additional_headers, retry_fn)
+  local request = Request.new("PUT", path, nil):add_header(
+                    "user-agent", "smartthings-lua-edge-driver"
+                  ):add_header("host", string.format("%s", self.base_url.host)):add_header(
+                    "connection", "keep-alive"
+                  )
+
+  if additional_headers ~= nil and type(additional_headers) == "table" then
+    for k, v in pairs(additional_headers) do request = request:add_header(k, v) end
+  end
+
+  request = request:append_body(body_string)
+
+  return execute_request(self, request, retry_fn)
+end
+
+function RestClient.new(base_url, sock_builder)
+  base_url = utils.force_url_table(base_url)
+
+  if type(sock_builder) ~= "function" then sock_builder = make_socket end
+
+  return
+    setmetatable({base_url = base_url, socket_builder = sock_builder, socket = nil}, RestClient)
+end
+
+return RestClient
```

Comment:
I would recommend a general code review of this module in Hue and Sonos again, as it has changed quite a bit in the last month or two as we identified issues after migration. Not all of the changes will apply here, but some of the changes that we made such as moving a lot of the work on to a separate cosock task have been beneficial for us and may be beneficial for this driver as well.

Reply:
Ok thanks.